src/assets/scripts/client/aircraft/FlightManagementSystem/Fms.js

```diff
@@ -0,0 +1,278 @@
+         */
+        this.category = '';
+
+        this.init(aircraftInitProps);
+    }
+
+    /**
+     *
+     *
+     * @property currentWaypoint
+     * @return {WaypointModel}
+     */
+    get currentWaypoint() {
+        return this.legCollection[0].currentWaypoint;
+    }
+
+    /**
+     *
+     *
+     * @property currentLeg
+     * @return {LegModel}
+     */
+    get currentLeg() {
+        return this.legCollection[0];
+    }
+
+    /**
+     *
+     * @property currentRoute
+     * @return {string}
+     */
+    get currentRoute() {
+        const routeSegments = _map(this.legCollection, (legModel) => legModel.routeString);
+
+        return routeSegments.join('..');
+    }
+
+    /**
+     *
+     *
+     * @for Fms
+     * @method init
+     * @param aircraftInitProps {object}
+     */
+    init(aircraftInitProps) {
+        this.category = aircraftInitProps.category;
+
+        this.legCollection = this._buildInitialLegCollection(aircraftInitProps);
+    }
+
+    /**
+     *
+     *
+     * @for LegModel
+     * @method destroy
+     */
+    destroy() {
+        this._navigationLibrary = null;
+        this._runway = '';
+        this.legCollection = [];
+        this.category = '';
+    }
+
+    /**
+     *
+     *
+     */
+    setModeControllerMode(modeSelector, mode) {
+        this._modeController[modeSelector] = mode;
+    }
+
+    /**
+     *
+     *
+     */
+    setModeControllerValue(fieldName, value) {
+        this._modeController[fieldName] = value;
+    }
+
+    updateModesForArrival() {
+        this._modeController.setForArrival();
+    }
+
+
+    updateModesForDeparture() {
+        this._modeController.setForDeparture();
+    }
+
+    /**
+     * Add a new `LegModel` to the left side of the `legCollection`
+     *
+     * @for Fms
+     * @method addLegToBeginning
+     * @param routeString
+     */
+    addLegToBeginning(routeString) {
+        const legModel = new LegModel(routeString, this._runway, this.category, this._navigationLibrary);
+
+        this.legCollection.unshift(legModel);
+    }
+
+    /**
+     * Move to the next possible waypoint
+     *
+     * This could be the next waypoint in the current leg,
+     * or the first waypoint in the next leg.
+     *
+     * @for LegModel
+     * @method nextWaypoint
+     */
+    nextWaypoint() {
+        if (!this.currentLeg.hasNextWaypoint()) {
+            this._moveToNextLeg();
+        }
+
+        this._moveToNextWaypointInLeg();
+    }
+
+
+    /**
+     * Given a `waypointName`, find where that waypoint exists within
+     * the `#legsColelction` then make that Leg active and `waypointName`
+     * the active waypoint for that Leg.
+     *
+     * @for Fms
+     * @method skipToWaypoint
+     * @param waypointName {string}
+     */
+    skipToWaypoint(waypointName) {
+        const { legIndex, waypointIndex } = this._findLegAndWaypointIndexForWaypointName(waypointName);
+
+        this.legCollection = _drop(this.legCollection, legIndex);
```

Comment:
>I'm more concerned with why you would need access to old portions of a route. what are some examples of why we would need to store an old part of a route?

Because you just _do_. From the air traffic control perspective, having passed a fix changes nothing about your route. If I ask the system what an aircraft's route is, I will get the whole thing, not just the remaining portion. From the aircraft's perspective, if it's in the past, it's gone.

Any time we expose the route to the controller, it should be shown in full, not in part. If there's no way to achieve this, we can break from reality here to avoid unnecessary effort, but the expectation would be that the controller would see the entire flightplan route. Other than visually comparing with where they see the aircraft on the screen, they wouldn't even have the ability to get the "remaining portion" of the route, so it would just feel really weird to see it that way. Especially if they arrive at the end of the STAR, and receive vectors. Their "route" shown to the controller would be... empty.